File: core/src/com/megaman/maverick/game/screens/menus/bosses/BossSelectScreen.java
package com.megaman.maverick.game.screens.menus.bosses;

import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.g2d.TextureAtlas;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.math.Vector3;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.ObjectMap;
import com.badlogic.gdx.utils.ObjectSet;
import com.engine.animations.Animation;
import com.engine.common.enums.Direction;
import com.engine.common.enums.Position;
import com.engine.common.time.TimeMarkedRunnable;
import com.engine.common.time.Timer;
import com.engine.drawables.fonts.BitmapFontHandle;
import com.engine.screens.menus.IMenuButton;
import com.megaman.maverick.game.ConstVals;
import com.megaman.maverick.game.MegamanMaverickGame;
import com.megaman.maverick.game.assets.MusicAsset;
import com.megaman.maverick.game.assets.SoundAsset;
import com.megaman.maverick.game.assets.TextureAsset;
import com.megaman.maverick.game.entities.bosses.BossType;
import com.megaman.maverick.game.screens.ScreenEnum;
import com.megaman.maverick.game.screens.menus.AbstractMenuScreen;
import com.megaman.maverick.game.screens.utils.BlinkingArrow;
import com.megaman.maverick.game.screens.utils.ScreenSlide;
import com.megaman.maverick.game.utils.ExtensionsKt;
import org.jetbrains.annotations.NotNull;

import java.util.EnumMap;
import java.util.Map;
import java.util.function.Supplier;

public class BossSelectScreen extends AbstractMenuScreen {

    private static final Vector3 INTRO_BLOCKS_TRANS = new Vector3(15f * ConstVals.PPM, 0f, 0f);
    private static final String MEGA_MAN = "MEGA MAN";
    private static final String BACK = "BACK";

    private static ObjectSet<String> bNameSet;

    private final BitmapFontHandle bName;
    private final ScreenSlide slide;
    private final Sprite bar1;
    private final Sprite bar2;
    private final Sprite white;
    private final Timer outTimer;
    private final Array<BitmapFontHandle> t;
    private final Array<BossPane> bp;
    private final Array<Sprite> bkgd;
    private final ObjectMap<Sprite, Animation> bars;
    private final ObjectMap<String, BlinkingArrow> bArrs;

    private boolean outro;
    private boolean blink;
    private BossType bSelect;

    private ObjectMap<String, IMenuButton> _menuButtons;

    public BossSelectScreen(MegamanMaverickGame game) {
        super(game, MEGA_MAN);
        if (bNameSet == null) {
            bNameSet = new ObjectSet<>();
            for (BossType b : BossType.values()) {
                bNameSet.add(b.name);
            }
        }
        bar1 = new Sprite();
        bar2 = new Sprite();
        white = new Sprite();
        t = new Array<>();
        bp = new Array<>();
        bkgd = new Array<>();
        bars = new ObjectMap<>();
        bArrs = new ObjectMap<>();
        outTimer = new Timer(1.05f, new Array<>() {{
            for (int i = 1; i <= 10; i++) {
                add(new TimeMarkedRunnable(.1f * i, () -> {
                    blink = !blink;
                    return null;
                }));
            }
        }});
        slide = new ScreenSlide(getCastGame().getUiCamera(), INTRO_BLOCKS_TRANS,
                ExtensionsKt.getDefaultCameraPosition().sub(INTRO_BLOCKS_TRANS),
                ExtensionsKt.getDefaultCameraPosition(), .5f, false);
        TextureAtlas megamanFacesAtlas = getCastGame().assMan.get(TextureAsset.FACES_1.getSource(), TextureAtlas.class);
        Map<Position, TextureRegion> megamanFaces = new EnumMap<>(Position.class);
        for (Position position : Position.values()) {

            // TODO: Maverick instead of Megaman faces?
            TextureRegion faceRegion = megamanFacesAtlas.findRegion("Maverick/" + position.name());
            megamanFaces.put(position, faceRegion);

        }
        Supplier<TextureRegion> megamanFaceSupplier = () -> {
            BossType boss = BossType.findByName(getCurrentButtonKey());
            if (boss == null) {
                return megamanFaces.get(Position.CENTER);
            }
            return megamanFaces.get(boss.position);
        };
        BossPane megamanPane = new BossPane(getCastGame(), megamanFaceSupplier, MEGA_MAN, Position.CENTER);
        bp.add(megamanPane);
        for (BossType boss : BossType.values()) {
            bp.add(new BossPane(getCastGame(), boss));
        }
        t.add(new BitmapFontHandle(() -> "PRESS START", ExtensionsKt.getDefaultFontSize(),
                new Vector2(5.35f * ConstVals.PPM, 13.85f * ConstVals.PPM), true, true, "Megaman10Font.ttf"));
        t.add(new BitmapFontHandle(() -> BACK, ExtensionsKt.getDefaultFontSize(), new Vector2(12.35f * ConstVals.PPM,
                ConstVals.PPM), true, true, "Megaman10Font.ttf"));
        bArrs.put(BACK, new BlinkingArrow(getCastGame().assMan, new Vector2(12f * ConstVals.PPM,
                .75f * ConstVals.PPM)));
        TextureAtlas stageSelectAtlas = getCastGame().assMan.get(TextureAsset.UI_1.getSource(), TextureAtlas.class);
        TextureRegion bar = stageSelectAtlas.findRegion("Bar");
        for (int i = 0; i < 6; i++) {
            for (int j = 0; j < 3; j++) {
                Sprite sprite = new Sprite(bar);
                sprite.setBounds(i * 3f * ConstVals.PPM, (j * 4f * ConstVals.PPM) + 1.35f * ConstVals.PPM,
                        5.33f * ConstVals.PPM, 4f * ConstVals.PPM);
                Animation timedAnimation = new Animation(bar, 1, 4, Array.with(.3f, .15f, .15f, .15f), true);
                bars.put(sprite, timedAnimation);
            }
        }
        TextureAtlas colorsAtlas = getCastGame().assMan.get(TextureAsset.COLORS.getSource(), TextureAtlas.class);
        TextureRegion whiteReg = colorsAtlas.findRegion("White");
        white.setRegion(whiteReg);
        white.setBounds(0f, 0f, ConstVals.VIEW_WIDTH * ConstVals.PPM, ConstVals.VIEW_HEIGHT * ConstVals.PPM);
        TextureRegion black = colorsAtlas.findRegion("Black");
        bar1.setRegion(black);
        bar1.setBounds(-ConstVals.PPM, -ConstVals.PPM, (2f + ConstVals.VIEW_WIDTH) * ConstVals.PPM, 2f * ConstVals.PPM);
        bar2.setRegion(black);
        bar2.setBounds(0f, 0f, .25f * ConstVals.PPM, ConstVals.VIEW_HEIGHT * ConstVals.PPM);
        TextureAtlas tilesAtlas = getCastGame().assMan.get(TextureAsset.PLATFORMS_1.getSource(), TextureAtlas.class);
        TextureRegion blueBlockRegion = tilesAtlas.findRegion("8bitBlueBlockTransBorder");
        final float halfPPM = ConstVals.PPM / 2f;
        for (int i = 0; i < ConstVals.VIEW_WIDTH; i++) {
            for (int j = 0; j < ConstVals.VIEW_HEIGHT - 1; j++) {
                for (int x = 0; x < 2; x++) {
                    for (int y = 0; y < 2; y++) {
                        Sprite blueBlock = new Sprite(blueBlockRegion);
                        blueBlock.setBounds(i * ConstVals.PPM + (x * halfPPM), j * ConstVals.PPM + (y * halfPPM),
                                halfPPM, halfPPM);
                        bkgd.add(blueBlock);
                    }
                }
            }
        }
        bName = new BitmapFontHandle(() -> "", ExtensionsKt.getDefaultFontSize(), new Vector2(ConstVals.PPM,
                ConstVals.PPM), true, true, "Megaman10Font.ttf");
    }

    @Override
    public void show() {
        super.show();
        slide.init();
        outro = false;
        outTimer.reset();
        getCastGame().getAudioMan().playMusic(MusicAsset.STAGE_SELECT_MM3_MUSIC, true);
    }

    @Override
    protected void onAnyMovement() {
        getCastGame().audioMan.playSound(SoundAsset.CURSOR_MOVE_BLOOP_SOUND, false);
    }

    @Override
    public void render(float delta) {
        super.render(delta);
        Batch batch = getCastGame().batch;
        if (!getCastGame().getPaused()) {
            slide.update(delta);
            if (outro) {
                outTimer.update(delta);
            }
            if (outTimer.isFinished()) {
                getCastGame().startLevelScreen(bSelect.level);
                return;
            }
            for (ObjectMap.Entry<Sprite, Animation> e : bars) {
                e.value.update(delta);
                e.key.setRegion(e.value.getCurrentRegion());
            }
            for (BossPane b : bp) {
                if (b.getBossName().equals(getCurrentButtonKey())) {
                    b.setBossPaneStat(getSelectionMade() ? BossPaneStat.HIGHLIGHTED : BossPaneStat.BLINKING);
                } else {
                    b.setBossPaneStat(BossPaneStat.UNHIGHLIGHTED);
                }
                b.update(delta);
            }
            if (bArrs.containsKey(getCurrentButtonKey())) {
                bArrs.get(getCurrentButtonKey()).update(delta);
            }
        }
        batch.setProjectionMatrix(getCastGame().getUiCamera().combined);
        batch.begin();
        if (outro && blink) {
            white.draw(batch);
        }
        for (Sprite b : bkgd) {
            b.draw(batch);
        }
        for (ObjectMap.Entry<Sprite, Animation> e : bars) {
            e.key.draw(batch);
        }
        for (BossPane b : bp) {
            b.draw(batch);
        }
        bar1.draw(batch);
        bar2.draw(batch);
        if (bArrs.containsKey(getCurrentButtonKey())) {
            bArrs.get(getCurrentButtonKey()).draw(batch);
        }
        for (BitmapFontHandle text : t) {
            text.draw(batch);
        }
        if (MEGA_MAN.equals(getCurrentButtonKey()) || bNameSet.contains(getCurrentButtonKey())) {
            bName.setTextSupplier(() -> getCurrentButtonKey().toUpperCase());
            bName.draw(batch);
        }
        batch.end();
    }

    @Override
    protected @NotNull ObjectMap<String, IMenuButton> getMenuButtons() {
        if (_menuButtons != null) {
            return _menuButtons;
        }

        _menuButtons = new ObjectMap<>();
        _menuButtons.put(MEGA_MAN, new IMenuButton() {
            @Override
            public boolean onSelect(float delta) {
                return false;
            }

            @Override
            public String onNavigate(@NotNull Direction direction, float delta) {
                return switch (direction) {
                    case UP -> BossType.findByPos(1, 2).name;
                    case DOWN -> BossType.findByPos(1, 0).name;
                    case LEFT -> BossType.findByPos(0, 1).name;
                    case RIGHT -> BossType.findByPos(2, 1).name;
                };
            }
        });
        _menuButtons.put(BACK, new IMenuButton() {
            @Override
            public boolean onSelect(float delta) {
                getCastGame().setCurrentScreen(ScreenEnum.MAIN.name());
                return true;
            }

            @Override
            public String onNavigate(@NotNull Direction direction, float delta) {
                return switch (direction) {
                    case UP, LEFT, RIGHT -> BossType.findByPos(2, 0).name;
                    case DOWN -> BossType.findByPos(2, 2).name;
                };
            }
        });
        for (BossType boss : BossType.values()) {
            _menuButtons.put(boss.name, new IMenuButton() {
                @Override
                public boolean onSelect(float delta) {
                    getCastGame().audioMan.playSound(SoundAsset.BEAM_OUT_SOUND, false);
                    getCastGame().audioMan.stopMusic(null);
                    bSelect = boss;
                    outro = true;
                    return true;
                }

                @Override
                public String onNavigate(@NotNull Direction direction, float delta) {
                    int x = boss.position.getX();
                    int y = boss.position.getY();
                    switch (direction) {
                        case UP -> y += 1;
                        case DOWN -> y -= 1;
                        case LEFT -> x -= 1;
                        case RIGHT -> x += 1;
                    }
                    if (y < 0 || y > 2) {
                        return BACK;
                    }
                    if (x < 0) {
                        x = 2;
                    }
                    if (x > 2) {
                        x = 0;
                    }
                    Position position = Position.Companion.get(x, y);
                    if (position == null) {
                        throw new IllegalStateException();
                    } else if (position.equals(Position.CENTER)) {
                        return MEGA_MAN;
                    } else {
                        return BossType.findByPos(x, y).name;
                    }
                }
            });
        }
        return _menuButtons;
    }

}

File: core/src/com/megaman/maverick/game/screens/menus/bosses/BossSelectScreen.kt
package com.megaman.maverick.game.screens.menus.bosses

import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.graphics.g2d.Sprite
import com.badlogic.gdx.graphics.g2d.TextureAtlas
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.math.Vector3
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.ObjectMap
import com.badlogic.gdx.utils.ObjectSet
import com.mega.game.engine.animations.Animation
import com.mega.game.engine.common.enums.Direction
import com.mega.game.engine.common.enums.Position
import com.mega.game.engine.common.enums.Position.Companion.get
import com.mega.game.engine.common.interfaces.Initializable
import com.mega.game.engine.common.time.TimeMarkedRunnable
import com.mega.game.engine.common.time.Timer
import com.mega.game.engine.drawables.fonts.BitmapFontHandle
import com.mega.game.engine.screens.menus.IMenuButton
import com.megaman.maverick.game.ConstVals
import com.megaman.maverick.game.MegamanMaverickGame
import com.megaman.maverick.game.assets.MusicAsset
import com.megaman.maverick.game.assets.SoundAsset
import com.megaman.maverick.game.assets.TextureAsset
import com.megaman.maverick.game.entities.bosses.BossType
import com.megaman.maverick.game.screens.ScreenEnum
import com.megaman.maverick.game.screens.menus.MegaMenuScreen
import com.megaman.maverick.game.screens.utils.BlinkingArrow
import com.megaman.maverick.game.screens.utils.ScreenSlide
import com.megaman.maverick.game.utils.MegaUtilMethods.getDefaultFontSize
import com.megaman.maverick.game.utils.getDefaultCameraPosition
import java.util.*
import java.util.function.Supplier

class BossSelectScreen(game: MegamanMaverickGame) : MegaMenuScreen(game, MEGA_MAN), Initializable {

    companion object {
        private val INTRO_BLOCKS_TRANS = Vector3(15f * ConstVals.PPM, 0f, 0f)
        private val CAM_POS = getDefaultCameraPosition().add(0f, 0.55f * ConstVals.PPM, 0f)
        private const val MEGA_MAN = "MEGA MAN"
        private const val BACK = "BACK"
    }

    private val bNameSet = ObjectSet<String>()
    private val slide = ScreenSlide(
        game.getUiCamera(),
        INTRO_BLOCKS_TRANS,
        CAM_POS.cpy().sub(INTRO_BLOCKS_TRANS),
        CAM_POS,
        0.5f,
        false
    )
    private val bar1 = Sprite()
    private val bar2 = Sprite()
    private val white = Sprite()
    private val outTimer = Timer(1.05f)
    private val t = Array<BitmapFontHandle>()
    private val bp = Array<BossPane>()
    private val bkgd = Array<Sprite>()
    private val bars = ObjectMap<Sprite, Animation>()
    private val bArrs = ObjectMap<String, BlinkingArrow>()
    private lateinit var bName: BitmapFontHandle
    private var outro = false
    private var blink = false
    private var bSelect: BossType? = null
    private var initialized = false

    override fun init() {
        if (initialized) return

        buttons.put(MEGA_MAN, object : IMenuButton {
            override fun onSelect(delta: Float) = false

            override fun onNavigate(direction: Direction, delta: Float): String? {
                return when (direction) {
                    Direction.UP -> BossType.findByPos(1, 2).name
                    Direction.DOWN -> BossType.findByPos(1, 0).name
                    Direction.LEFT -> BossType.findByPos(0, 1).name
                    Direction.RIGHT -> BossType.findByPos(2, 1).name
                }
            }
        })
        buttons.put(BACK, object : IMenuButton {
            override fun onSelect(delta: Float): Boolean {
                game.setCurrentScreen(ScreenEnum.MAIN_MENU_SCREEN.name)
                return true
            }

            override fun onNavigate(direction: Direction, delta: Float): String? {
                return when (direction) {
                    Direction.UP, Direction.LEFT, Direction.RIGHT -> BossType.findByPos(2, 0).name
                    Direction.DOWN -> BossType.findByPos(2, 2).name
                }
            }
        })
        for (boss in BossType.values()) {
            buttons.put(boss.name, object : IMenuButton {
                override fun onSelect(delta: Float): Boolean {
                    game.audioMan.playSound(SoundAsset.BEAM_OUT_SOUND, false)
                    game.audioMan.stopMusic(null)
                    bSelect = boss
                    outro = true
                    return true
                }

                override fun onNavigate(direction: Direction, delta: Float): String? {
                    var x = boss.position.x
                    var y = boss.position.y
                    when (direction) {
                        Direction.UP -> y += 1
                        Direction.DOWN -> y -= 1
                        Direction.LEFT -> x -= 1
                        Direction.RIGHT -> x += 1
                    }
                    if (y < 0 || y > 2) return BACK
                    if (x < 0) x = 2
                    if (x > 2) x = 0
                    val position = get(x, y)
                    return when (position) {
                        null -> throw IllegalStateException()
                        Position.CENTER -> MEGA_MAN
                        else -> BossType.findByPos(x, y).name
                    }
                }
            })
        }

        for (b in BossType.values()) bNameSet.add(b.name)
        val outTimerRunnable = Array<TimeMarkedRunnable>()
        for (i in 1..10) outTimerRunnable.add(TimeMarkedRunnable(.1f * i) { blink = !blink })
        outTimer.setRunnables(outTimerRunnable)
        val megamanFacesAtlas = game.assMan.get(TextureAsset.FACES_1.source, TextureAtlas::class.java)
        val megamanFaces: MutableMap<Position, TextureRegion> = EnumMap(Position::class.java)
        for (position in Position.values()) {
            val faceRegion: TextureRegion = megamanFacesAtlas.findRegion("Megaman/" + position.name)
            megamanFaces[position] = faceRegion
        }
        val megamanFaceSupplier = Supplier {
            val boss = BossType.findByName(currentButtonKey) ?: return@Supplier megamanFaces[Position.CENTER]
            megamanFaces[boss.position]
        }
        val megamanPane = BossPane(game, megamanFaceSupplier, MEGA_MAN, Position.CENTER)
        bp.add(megamanPane)
        for (boss in BossType.values()) bp.add(BossPane(game, boss))
        t.add(
            BitmapFontHandle(
                "PRESS START",
                getDefaultFontSize(),
                Vector2(5.35f * ConstVals.PPM, 13.85f * ConstVals.PPM),
                centerX = false,
                centerY = false,
                fontSource = "Megaman10Font.ttf"
            )
        )
        t.add(
            BitmapFontHandle(
                BACK, getDefaultFontSize(), Vector2(
                    12.35f * ConstVals.PPM, ConstVals.PPM.toFloat()
                ), false, centerY = false, fontSource = "Megaman10Font.ttf"
            )
        )
        bArrs.put(
            BACK, BlinkingArrow(
                game.assMan, Vector2(
                    12f * ConstVals.PPM, .75f * ConstVals.PPM
                )
            )
        )
        val stageSelectAtlas = game.assMan.get(TextureAsset.UI_1.source, TextureAtlas::class.java)
        val bar: TextureRegion = stageSelectAtlas.findRegion("Bar")
        for (i in 0..5) {
            for (j in 0..2) {
                val sprite = Sprite(bar)
                sprite.setBounds(
                    i * 3f * ConstVals.PPM,
                    (j * 4f * ConstVals.PPM) + 1.35f * ConstVals.PPM,
                    5.33f * ConstVals.PPM,
                    4f * ConstVals.PPM
                )
                val timedAnimation = Animation(bar, 1, 4, Array.with(.3f, .15f, .15f, .15f), true)
                bars.put(sprite, timedAnimation)
            }
        }
        val colorsAtlas = game.assMan.get(TextureAsset.COLORS.source, TextureAtlas::class.java)
        val whiteReg: TextureRegion = colorsAtlas.findRegion("White")
        white.setRegion(whiteReg)
        white.setBounds(0f, 0f, ConstVals.VIEW_WIDTH * ConstVals.PPM, ConstVals.VIEW_HEIGHT * ConstVals.PPM)
        val black: TextureRegion = colorsAtlas.findRegion("Black")
        bar1.setRegion(black)
        bar1.setBounds(
            -ConstVals.PPM.toFloat(),
            -ConstVals.PPM.toFloat(),
            (2f + ConstVals.VIEW_WIDTH) * ConstVals.PPM,
            2f * ConstVals.PPM
        )
        bar2.setRegion(black)
        bar2.setBounds(0f, 0f, 0.25f * ConstVals.PPM, (ConstVals.VIEW_HEIGHT + 1) * ConstVals.PPM)
        val tilesAtlas = game.assMan.get(TextureAsset.PLATFORMS_1.source, TextureAtlas::class.java)
        val blueBlockRegion: TextureRegion = tilesAtlas.findRegion("8bitBlueBlockTransBorder")
        val halfPPM = ConstVals.PPM / 2f
        var i = 0
        while (i < ConstVals.VIEW_WIDTH) {
            var j = 0
            while (j < ConstVals.VIEW_HEIGHT - 1) {
                for (x in 0..1) {
                    for (y in 0..1) {
                        val blueBlock = Sprite(blueBlockRegion)
                        blueBlock.setBounds(
                            i * ConstVals.PPM + (x * halfPPM), j * ConstVals.PPM + (y * halfPPM), halfPPM, halfPPM
                        )
                        bkgd.add(blueBlock)
                    }
                }
                j++
            }
            i++
        }
        bName = BitmapFontHandle(
            { "" }, getDefaultFontSize(), Vector2(
                ConstVals.PPM.toFloat(), ConstVals.PPM.toFloat()
            ), false, centerY = false, fontSource = "Megaman10Font.ttf"
        )
    }

    override fun show() {
        if (!initialized) {
            init()
            initialized = true
        }
        super.show()
        slide.init()
        outro = false
        outTimer.reset()
        game.audioMan.playMusic(MusicAsset.MM3_SNAKE_MAN_MUSIC, true)
    }

    override fun onAnyMovement(direction: Direction) =
        game.audioMan.playSound(SoundAsset.CURSOR_MOVE_BLOOP_SOUND, false)

    override fun render(delta: Float) {
        super.render(delta)
        val batch: Batch = game.batch
        if (!game.paused) {
            slide.update(delta)
            if (outro) outTimer.update(delta)
            if (outTimer.isFinished()) {
                /*
                TODO:
                val bIntroScreen = game.screens.get(ScreenEnum.BOSS_INTRO_SCREEN.name) as BossIntroScreen
                bIntroScreen.set(bSelect!!)
                game.setCurrentScreen(ScreenEnum.BOSS_INTRO_SCREEN.name)
                 */
                game.startLevelScreen(bSelect!!.level)
                return
            }
            for (e in bars) {
                e.value.update(delta)
                e.key.setRegion(e.value.getCurrentRegion())
            }
            for (b in bp) {
                b.bossPaneStat =
                    if (b.bossName == currentButtonKey) (if (selectionMade) BossPaneStat.HIGHLIGHTED else BossPaneStat.BLINKING)
                    else BossPaneStat.UNHIGHLIGHTED
                b.update(delta)
            }
            if (bArrs.containsKey(currentButtonKey)) bArrs.get(currentButtonKey).update(delta)
        }
        batch.projectionMatrix = game.getUiCamera().combined
        batch.begin()
        if (outro && blink) white.draw(batch)
        for (b in bkgd) b.draw(batch)
        for (e in bars) e.key.draw(batch)
        for (b in bp) b.draw(batch)
        bar1.draw(batch)
        bar2.draw(batch)
        if (bArrs.containsKey(currentButtonKey)) bArrs.get(currentButtonKey).draw(batch)
        for (text in t) text.draw(batch)
        if (MEGA_MAN == currentButtonKey || bNameSet.contains(currentButtonKey)) {
            bName.textSupplier = { currentButtonKey!!.uppercase(Locale.getDefault()) }
            bName.draw(batch)
        }
        batch.end()
    }
}

